benchalerts/parse_conbench.py

```diff
@@ -12,26 +12,116 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import textwrap
 from typing import List, Tuple
 
+from .clients import GitHubRepoClient
+
+
+def _clean(text: str) -> str:
+    """Clean text so it displays nicely as GitHub Markdown."""
+    return textwrap.fill(textwrap.dedent(text), 10000).replace("  ", "\n\n").strip()
+
 
 def benchmarks_with_z_regressions(all_comparisons: dict) -> List[Tuple[str, str]]:
-    """Find the run_ids and names of benchmarks whose z-scores were extreme enough to
-    constitute a regression.
+    """Find the compare_urls and names of benchmarks whose z-scores were extreme enough
+    to constitute a regression.
 
     Parameters
     ----------
     all_comparisons
-        The output of ConbenchClient.get_comparison_to_baseline().
+        The dict output of ConbenchClient.get_comparison_to_baseline().
 
     Returns
     -------
     List[Tuple[str, str]]
-        List of tuples of (run_id, benchmark_name) for all benchmarks with regressions.
+        List of tuples of (compare_url, benchmark_name) for all benchmarks with
+        regressions.
     """
     return [
-        (run_id, benchmark["benchmark"])
-        for run_id, benchmarks in all_comparisons.items()
+        (compare_url, benchmark["benchmark"])
+        for compare_url, benchmarks in all_comparisons.items()
         for benchmark in benchmarks
         if benchmark["contender_z_regression"]
     ]
+
+
+def regression_summary(
+    all_comparisons: dict,
+    baseline_is_parent: bool,
+    contender_sha: str,
+    warn_if_baseline_isnt_parent: bool,
+) -> str:
+    """Generate a Markdown summary of what happened regarding regressions."""
+    if not all_comparisons:
+        return _clean(
+            f"""
+            Conbench could not find a baseline run for contender commit
+            `{contender_sha[:8]}`. A baseline run needs to be on the default branch,
+            with the same hardware, repository, case, and context as the contender run.
```

Comment:
I'll admit I don't have nearly enough knowledge on this as I should. I was going off of [this code](https://github.com/conbench/conbench/blob/main/conbench/entities/run.py#L68-L71) which seems to be only returning a parent run if it matches the `case_id` and `context_id` of the contender run. But I could be misunderstanding.

I'm happy to adjust this to be more correct! Do you or @jonkeane have a suggestion?

Reply:
oh man so apparently we have cases on the client side and [the server side](https://github.com/conbench/conbench/blob/main/conbench/entities/benchmark_result.py#L96-L99) but not in the API in the middle 😫

Reply:
> oh man so apparently we have cases on the client side and [the server side](https://github.com/conbench/conbench/blob/main/conbench/entities/benchmark_result.py#L96-L99) but not in the API in the middle 😫 

We could change this if that helps, of course

As for the actual question: IMO I would say something like "A baseline run needs to be on the default branch in the same repository, with the same hardware and context, and have at least one of the same benchmarks"

This a bit sidesteps the concept of cases vs. benchmarks, but I suspect in 90% of parlance about benchmarks people think of each case as a separate benchmark (especially in this context)

Reply:
We've elsewhere defined "benchmark" as the code the collection of cases runs on

Reply:
Cool, I put exactly what you wrote but changed the last part to "at least one of the same benchmark cases." I think that's clear enough.

Reply:
> We've elsewhere defined "benchmark" as the code the collection of cases runs on

Did we degine that as "The code the collection of cases runs on [with the parameters of the cases filled in]" or "The code the collection of cases runs on [with the parameters not-filled]"? 😛 

Obviously not in this PR, but we should probably write some details in docs about this distinction and think about if we should either go against it (find new name(s) for these things and use them relentlessly until people follow) or if this is a place where we can lean into the ambiguity for the most part and only go into the pedantic distinction when it's necessary (i.e. the conbench API, or for folks authoring benchmarks). I don't know which of those is the right approach (or if there's another third way...)

Reply:
On the client side, `Benchmark` is a class (well, with like 7 versions in 2 languages, but fewer eventually) that holds the code to run and a default set of cases to run it with. Running an instance with defaults will run multiple cases and usually return a list of `BenchmarkResult` instances.

What we're talking about here is at the `BenchmarkResult` level, though it is admittedly confusing to users of the website that one `Benchmark` may have lots of `BenchmarkResult`s (because cases), so maybe we need more words. Maybe `BenchmarkResult` should become `BenchmarkCaseResult` or something?

Reply:
Made https://github.com/conbench/conbench/issues/401